### PR TITLE
Ross Wollman

### PR DESCRIPTION
**Pull request by rwoll-hmc**

### CODE REVIEW

File: context-free.md [hunk 1]
```diff
@@ -2,21 +2,65 @@
...
 
+It is easy to make pattern filled, recursive, and random images due to the
+nature of the language. The loop syntax is especially helpful in making
+interesting patterns.
+
+It is also easy, no matter the end picture, to understand the composobility of
+shapes if the code chunks are factored appropriately to different "shape" names.
+Since we can nest primitive and custom shapes within one another, it makes it
+easier to preciesly refer to a part of the picture should someone need to give
+feedback or design critiques.
```
Comment (hmc-cs-jbearer) [inline context-free.md:20]: This is an interesting point. I definitely agree, although I hadn't thought about this before. For instance, looking at your ContextFree code actually helped me appreciate the resulting image more, because I was able to pick apart the hierarchical levels of detail and structure in the code and then find those same structures in the image, which is cool

File: context-free.md [hunk 2]
```diff
@@ -2,21 +2,65 @@
...
 
 ## What is hard to do in this language? Why is it hard?
 
+It is hard to create pictures of non-organic items (or perhaps my artistic
+skills and familiarity with the language need more developing). Since
+ContextFree is code that starts with primitive shapes, one must try really hard
+to mentally divide a non-organic shape into primitive shapes if they want to
+draw it accurately. Making adjustments requires a bit of guess and check then
+a slow re-render which makes it harder.
```
Comment (hmc-cs-jbearer) [inline context-free.md:30]: All true. I think it's also important to recognize that it would be hard to do _anything_ other than the things you said were easy. I believe it's important to mention this, because you could imagine a language in which it is easy to produce lots of images, including "non-organic items", and this language could still be a DSL. It's the inability to do anything else that I believe is important for DSL-ness.

File: context-free.md [hunk 3]
```diff
@@ -2,21 +2,65 @@
...
+_We don't yet have a great definition of the term "computational model".
+For now, try to come up with the clearest, most concise explanation of what
 happens when a ContextFree program runs._
 
+At first, the parser or image builder looks for the top-level shape and then
+looks to see what other shapes it is composed of. Once it finally hits a
+primitive shape, it renders a unit of the larger shape with the specified
+characteristics. As it continues and the sub-shapes are drawn, the larger shapes
+naturally come are created. When there are no more loops to evaluate or
+primitives to draw it finishes.
```
Comment (hmc-cs-jbearer) [inline context-free.md:54]: It can also finish when any shapes yet to be drawn are below a certain size, although I'm not sure how it determines that size.

File: context-free.md [hunk 4]
```diff
@@ -2,21 +2,65 @@
...
+
+Since there is randomness embedded into the renderer, there is almost certainly
+a step before the rendering where the builder chooses a seed value for the
+"random" selections.
```
Comment (hmc-cs-jbearer) [inline context-free.md:59]: One part of the process that I think is missing here (although you hinted at it by talking about randomness) is rule selection, the basic manifestation of that randomness in a ContextFree program. It seems like, if there is more than one rule for how to draw a shape, the runtime will (pseudo) randomly choose one, with probability proportional to the weight given by the programmer.

File: my-dsl.md [hunk 5]
```diff
@@ -1,38 +1,105 @@
...
+```
+SEED data destination
+```
+
+which has a more human-friendly style to it.
+
+> Although I make the suggestion that Docker adds custom commands, it could
+> have the potential to make the Dockerfile Command Langauge too generic and
+> bash-like which Fowler would likely say is a bad thing for something that
+> claims to be a DSL.
```
Comment (hmc-cs-jbearer) [inline my-dsl.md:105]: I think I agree with this last point -- adding the ability to create custom commands sounds a lot like the feature creep we were talking about in class today. If the lack of custom commands is really a problem, though, it does beg the question whether a DSL is the right answer here. Would a Turing-complete Bash-like language get the job done more effectively, without the need for much scavenger-hunting?

I guess it probably comes down to tradeoffs: to get the readability and compartmentalization of Dockerfiles, you have to sacrifice a little bit of power. Hopefully as the semester goes on we'll start to develop an idea of when to use a DSL and when to use a more general tool.

File: my-dsl.md [hunk 6]
```diff
@@ -1,38 +1,105 @@
...
+looks a bit like a scripting language, the commands are extremely limited for
+the purposes of readability, reproducibility, and commitability. The langauge
+itself is certainly not Turing-complete&mdash;Fowler's litmus test for generic
+DSLs.
+
+While I definitely think the Dockerfile's commands are a DSL&mdash;the
+documentation even goes as far as refering to it as a DSL&mdash;it verges on the
+line of simply being a configuration file which Fowler would likely not count
+as a DSL. The reason I'd argue it is not simply a configuration file is there
+needs to be a logical ordering of steps that are user determined.
```
Comment (hmc-cs-jbearer) [inline my-dsl.md:45]: How does the ordering make Dockerfile a DSL? Maybe I'm misunderstanding, but I don't really see how this relates to our definition. Maybe the definition needs to be expanded?

File: context-free.md (hunk 3 above)
Comment (rwoll) [reply context-free.md:54]: Good point! My picture lacks recursion, so I excluded that, but it is definitely a fundamental property of the model.

File: context-free.md (hunk 4 above)
Comment (rwoll) [reply context-free.md:59]: Yes&mdash;thanks for expanding that point further.

File: context-free.md (hunk 2 above)
Comment (rwoll-hmc) [reply context-free.md:30]: I think there is definitely a relationship between limited expressiveness and DSL-ness, but I could imagine a point where such limited expressiveness makes the language no longer a language&mdash;let alone a DSL. At this point, it would be more like a configuration file.

tl;dr I think there is a **limited** relationship between DSL-ness and the limit of the language.

File: my-dsl.md (hunk 6 above)
Comment (rwoll-hmc) [reply my-dsl.md:45]: I didn't mean to imply that its ordering alone makes it a DSL, but I think it is a differentiating factor between a Dockerfile set of commands and a simple configuration file. In my eyes, it's just one test (of many) to distinguish the two.

File: my-dsl.md [hunk 7]
```diff
@@ -1,38 +1,105 @@
...
+A user describes the overall build of a container&mdash;a sort of lightweight
+virtual appliance&mdash;in a text file called a Dockerfile with limited text
+commands. When a user builds an image from the Dockerfile, the text commands are
+parsed and with each step a partial container is built representing the state of
+the machine&mdash;the Docker commands are converted to lower-level system
+commands as needed. After the final command, the user can use the built machine
+just as they would a virtual appliance without worrying about the intermediate
+steps. If the Dockerfile is edited, the image builder will find the modified
+line, and only rebuild and commit images that came from commands with and after
+that line.
```
Comment (hmc-cs-dobermiller) [inline my-dsl.md:26]: I'm not familiar with Docker (on the todo list...), but this sounds like a glorified config file with some Makefile thrown in. As such, it seems like it would be just barely a DSL - would you agree, or am I missing something?

File: my-dsl.md [hunk 8]
```diff
@@ -1,38 +1,105 @@
...
+The Dockerfile Command Langauge is "purely" domain-specific. It serves one
+purpose: to describe and build reproducible Docker containers/images. While it
+looks a bit like a scripting language, the commands are extremely limited for
+the purposes of readability, reproducibility, and commitability. The langauge
+itself is certainly not Turing-complete&mdash;Fowler's litmus test for generic
+DSLs.
+
+While I definitely think the Dockerfile's commands are a DSL&mdash;the
+documentation even goes as far as refering to it as a DSL&mdash;it verges on the
+line of simply being a configuration file which Fowler would likely not count
```
Comment (hmc-cs-dobermiller) [inline my-dsl.md:42]: Ah, I think this answers my question :smiley:

File: my-dsl.md (hunk 5 above)
Comment (hmc-cs-dobermiller) [reply my-dsl.md:105]: The ability to create custom commands makes it a lot easier to design for extensibility. It's certainly a tradeoff, and examining that tradeoff helps understand the intended audience of the DSL and the problem it is trying to solve.

In general, I personally prefer extensibility, but it's always a case-by-base thing.

File: ross_wollman_jfz.cfdg [hunk 9]
```diff
@@ -0,0 +1,19 @@
+startshape colorglobe
```
Comment (hmc-cs-dobermiller) [inline ross_wollman_jfz.cfdg:1]: This looks super neat!

**Review by hmc-cs-dobermiller — COMMENTED** [review]
Great work! You definitely dug into some of the core concepts while analyzing the two languages. Just keep looking at those topics and you'll do fine!
